### PR TITLE
Use ML-EID address in CoAP communication

### PR DESCRIPTION
**Pull request by lmaciejonczyk**
It fixes https://projecttools.nordicsemi.no/jira/browse/KRKNWK-6358

### CODE REVIEW

File: samples/openthread/coap_client/src/coap_client_utils.c [hunk 1]
```diff
@@ -51,6 +55,15 @@ static struct sockaddr_in6 unique_local_addr = {
...
 };
 
+/* OT ML-EID coap_client address is used as source address in communication 
+   with coap_server */
+static struct sockaddr_in6 coap_client_addr = {
+		.sin6_family = AF_INET6,
+		.sin6_port = htons(COAP_PORT),
+		.sin6_addr.s6_addr = {0, },
+		.sin6_scope_id = 0U,
+	};
```
Comment (rlubos) [inline samples/openthread/coap_client/src/coap_client_utils.c:65]: nit: indentation

File: subsys/net/lib/coap_utils/coap_utils.c [hunk 2]
```diff
@@ -49,6 +51,43 @@ static void coap_close_socket(int socket)
...
+	size_t addr_str_size;
+	const void *addr_src;
+
+	__ASSERT(addr != NULL, "Pointer is null");
+
+	switch (addr->sa_family) {
+	case AF_INET: {
+		socklen = sizeof(struct sockaddr_in);
+		addr_str_size = INET_ADDRSTRLEN;
+		addr_src = net_sin6(addr)->sin6_addr.s6_addr;
```
Comment (rlubos) [inline subsys/net/lib/coap_utils/coap_utils.c:67]: This should use IPv4 structures.

File: subsys/net/lib/coap_utils/coap_utils.c [hunk 3]
```diff
@@ -49,6 +51,43 @@ static void coap_close_socket(int socket)
...
+	} break;
+
+	case AF_INET6: {
+		socklen = sizeof(struct sockaddr_in6);
+		addr_str_size = INET6_ADDRSTRLEN;
+		addr_src = net_sin(addr)->sin_addr.s4_addr;
```
Comment (rlubos) [inline subsys/net/lib/coap_utils/coap_utils.c:73]: This should use IPv6 structures.

**Comment by lmaciejonczyk** [comment]
> This will need some investigation in Zephyr, as it has source matching algorithm implemented, so situation like this should not take place (i. e. mesh-local address should be used for on-mesh communication):
> https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/ip/net_if.c#L2327
> So there's either a bug in the algorithm, or the socket is implicitly bound to the link-local address (which is also bad).
> 
> Overall, this looks like a good solution.

Indeed there is a bug in the algorithm. Hopefully it will be fixed with this:
https://github.com/zephyrproject-rtos/zephyr/pull/26191

**Comment by lmaciejonczyk** [comment]
Fixed by: https://github.com/zephyrproject-rtos/zephyr/pull/26191

**Pull request closed without merging**